crates/wasi-common/src/snapshots/wasi_snapshot_preview1.rs

```diff
@@ -0,0 +1,1448 @@
+            fs_rights_base,
+            fs_rights_inheriting,
+            fdflags,
+        );
+
+        // Extract path as &str.
+        let mut bc = GuestBorrows::new();
+        let path = unsafe { &*path.as_raw(&mut bc)? };
+
+        trace!("     | path='{}'", path);
+
+        let (needed_base, needed_inheriting) =
+            path::open_rights(fs_rights_base, fs_rights_inheriting, oflags, fdflags);
+
+        trace!(
+            "     | needed_base = {}, needed_inheriting = {}",
+            needed_base,
+            needed_inheriting
+        );
+
+        let resolved = {
+            let entry = unsafe { self.get_entry(dirfd)? };
+            path::get(
+                &entry,
+                needed_base,
+                needed_inheriting,
+                dirflags,
+                path,
+                oflags & types::Oflags::CREAT != types::Oflags::empty(),
+            )?
+        };
+
+        // which open mode do we need?
+        let read = fs_rights_base & (types::Rights::FD_READ | types::Rights::FD_READDIR)
+            != types::Rights::empty();
+        let write = fs_rights_base
+            & (types::Rights::FD_DATASYNC
+                | types::Rights::FD_WRITE
+                | types::Rights::FD_ALLOCATE
+                | types::Rights::FD_FILESTAT_SET_SIZE)
+            != types::Rights::empty();
+
+        trace!(
+            "     | calling path_open impl: read={}, write={}",
+            read,
+            write
+        );
+
+        let fd = resolved.open_with(read, write, oflags, fdflags)?;
+        let mut fe = Entry::from(fd)?;
+        // We need to manually deny the rights which are not explicitly requested
+        // because FdEntry::from will assign maximal consistent rights.
+        fe.rights_base &= fs_rights_base;
+        fe.rights_inheriting &= fs_rights_inheriting;
+        let guest_fd = self.insert_entry(fe)?;
+
+        trace!("     | *fd={:?}", guest_fd);
+
+        Ok(guest_fd)
+    }
+
+    fn path_readlink(
+        &self,
+        dirfd: types::Fd,
+        path: &GuestPtr<'_, str>,
+        buf: GuestPtr<u8>,
+        buf_len: types::Size,
+    ) -> Result<types::Size> {
+        trace!(
+            "path_readlink(dirfd={:?}, path={:?}, buf={:?}, buf_len={})",
+            dirfd,
+            path,
+            buf,
+            buf_len,
+        );
+
+        let mut bc = GuestBorrows::new();
+        let as_str = path.as_raw(&mut bc)?;
+        let path = unsafe { &*as_str };
+
+        trace!("     | (path_ptr,path_len)='{}'", path);
+
+        let entry = unsafe { self.get_entry(dirfd)? };
+        let resolved = path::get(
+            &entry,
+            types::Rights::PATH_READLINK,
+            types::Rights::empty(),
+            types::Lookupflags::empty(),
+            path,
+            false,
+        )?;
+
+        // This is to account for `path_readlink` call with zero-length
+        // buffer. This is a valid POSIX call hence we need to accommodate.
```

Comment:
Hm does `as_raw` or `as_array` not work with zero-length arrays? I would naively expect this condition to not be necessary.

Reply:
You're right, I'd expect this to work as well. It turns out `wiggle_runtime::Region` didn't accommodate for zero-length `Region`s. I've now fixed that in #1366. I'll leave a comment in to remove the check once it lands in `wiggle-runtime`.

Reply:
Ok, since #1366 landed, I've now removed the check and everything works as expected :+1: